### PR TITLE
Fix STA apartment handling

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

File: src/coreclr/src/nativeaot/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs [hunk 1]
```diff
@@ -22,23 +22,32 @@ private static unsafe int WaitForMultipleObjectsIgnoringSyncContext(IntPtr* pHan
...
+                //
+                // In the CLR, we use CoWaitForMultipleHandles to pump messages while waiting in an STA.  In that case, we cannot use WAIT_ALL.
+                // That's because the wait would only be satisfied if a message arrives while the handles are signalled.
+                //
+                if (waitAll)
                     throw new NotSupportedException(SR.NotSupported_WaitAllSTAThread);
+
+                // CoWaitForMultipleHandles does not support more than 63 handles. It returns RPC_S_CALLPENDING for more than 63 handles
+                // that is impossible to differentiate from timeout.
+                if (numHandles > 63)
```
Comment (jkotas) [inline src/coreclr/src/nativeaot/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs:42]: This is fixing a test failure that results into fatal unhandled in System.Runtime.Tests run

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/IL/Stubs/StartupCode/StartupCodeMainMethod.cs [hunk 2]
```diff
@@ -92,6 +92,22 @@ public override MethodIL EmitIL()
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(initEntryAssembly));
             }
 
+            // Initialize COM apartment
+            MethodDesc initApartmentState = startup.GetMethod("InitializeApartmentState", null);
```
Comment (jkotas) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/IL/Stubs/StartupCode/StartupCodeMainMethod.cs:96]: Fixed this while I was on it.

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/IL/Stubs/StartupCode/StartupCodeMainMethod.cs [hunk 3]
```diff
@@ -92,6 +92,22 @@ public override MethodIL EmitIL()
...
+            if (initApartmentState != null)
+            {
+                if (_mainMethod.HasCustomAttribute("System", "STAThreadAttribute"))
+                {
+                    codeStream.EmitLdc(0 /* ApartmentState.STA */);
+                    codeStream.Emit(ILOpcode.call, emitter.NewToken(initApartmentState));
+                }
+                if (_mainMethod.HasCustomAttribute("System", "MTAThreadAttribute"))
```
Comment (jkotas) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/IL/Stubs/StartupCode/StartupCodeMainMethod.cs:104]: This structure is conveniently going to throw exception at runtime if one gets both STAThread and MTAThread attributes specified.

File: src/coreclr/src/nativeaot/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/src/nativeaot/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs:42]: Wouldn't even think to look in this direction. Thanks!

File: src/coreclr/src/tools/aot/ILCompiler.Compiler/IL/Stubs/StartupCode/StartupCodeMainMethod.cs [hunk 4]
```diff
@@ -92,6 +92,22 @@ public override MethodIL EmitIL()
                 codeStream.Emit(ILOpcode.call, emitter.NewToken(initEntryAssembly));
             }
 
+            // Initialize COM apartment
+            MethodDesc initApartmentState = startup.GetMethod("InitializeApartmentState", null);
+            if (initApartmentState != null)
+            {
+                if (_mainMethod.HasCustomAttribute("System", "STAThreadAttribute"))
+                {
+                    codeStream.EmitLdc(0 /* ApartmentState.STA */);
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.Compiler/IL/Stubs/StartupCode/StartupCodeMainMethod.cs:101]: Nit: We could probably just use the enum directly. The enums are guaranteed to not go out of sync.

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

File: src/coreclr/src/nativeaot/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs (hunk 1 above)
Comment (jkotas) [reply src/coreclr/src/nativeaot/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs:42]: It is not fixing the GC crash. It is fixing a different crash that was being hit some of the time.

**Pull request merged**